File: FBControlCore/Utility/FBControlCoreGlobalConfiguration.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBControlCoreGlobalConfiguration.h"

#import <Foundation/Foundation.h>

#import "FBControlCoreLogger.h"

NSString *const FBControlCoreStderrLogging = @"FBCONTROLCORE_LOGGING";
NSString *const FBControlCoreDebugLogging = @"FBCONTROLCORE_DEBUG_LOGGING";
NSString *const ConfirmShimsAreSignedEnv = @"FBCONTROLCORE_CONFIRM_SIGNED_SHIMS";

static id<FBControlCoreLogger> logger;

@implementation FBControlCoreGlobalConfiguration

+ (NSTimeInterval)fastTimeout
{
  return 10;
}

+ (NSTimeInterval)regularTimeout
{
  return 30;
}

+ (NSTimeInterval)slowTimeout
{
  return 120;
}

+ (id<FBControlCoreLogger>)defaultLogger
{
  if (logger) {
    return logger;
  }
  logger = [self createDefaultLogger];
  return logger;
}

+ (void)setDefaultLogger:(id<FBControlCoreLogger>)defaultLogger
{
  if (logger) {
    [defaultLogger.debug logFormat:@"Overriding the Default Logger with %@", defaultLogger];
  }
  logger = defaultLogger;
}

+ (BOOL)confirmCodesignaturesAreValid
{
  return NSProcessInfo.processInfo.environment[ConfirmShimsAreSignedEnv].boolValue;
}

+ (NSString *)description
{
  return [NSString stringWithFormat:@"Default Logger %@", logger];
}

- (NSString *)description
{
  return [FBControlCoreGlobalConfiguration description];
}

#pragma mark FBJSONConversion

- (id)jsonSerializableRepresentation
{
  return @{};
}

#pragma mark Private

+ (id<FBControlCoreLogger>)createDefaultLogger
{
  return [FBControlCoreLogger systemLoggerWritingToStderr:self.stderrLoggingEnabledByDefault withDebugLogging:self.debugLoggingEnabledByDefault];
}

+ (BOOL)stderrLoggingEnabledByDefault
{
  return [NSProcessInfo.processInfo.environment[FBControlCoreStderrLogging] boolValue];
}

+ (BOOL)debugLoggingEnabledByDefault
{
  return [NSProcessInfo.processInfo.environment[FBControlCoreDebugLogging] boolValue];
}

+ (nullable id)readValueForKey:(NSString *)key fromPlistAtPath:(NSString *)plistPath
{
  NSCAssert([NSFileManager.defaultManager fileExistsAtPath:plistPath], @"plist does not exist at path '%@'", plistPath);
  NSDictionary *infoPlist = [NSDictionary dictionaryWithContentsOfFile:plistPath];
  NSCAssert(infoPlist, @"Could not read plist at '%@'", plistPath);
  id value = infoPlist[key];
  NSCAssert(value, @"'%@' does not exist in plist '%@'", key, infoPlist.allKeys);
  return value;
}

@end

File: FBControlCore/Utility/FBControlCoreGlobalConfiguration.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

/// Environment variable; when truthy, the default logger mirrors its output to stderr.
public let FBControlCoreStderrLogging = "FBCONTROLCORE_LOGGING"
/// Environment variable; when truthy, the default logger emits debug-level output.
public let FBControlCoreDebugLogging = "FBCONTROLCORE_DEBUG_LOGGING"

private let ConfirmShimsAreSignedEnv = "FBCONTROLCORE_CONFIRM_SIGNED_SHIMS"

@objc(FBControlCoreGlobalConfiguration)
public class FBControlCoreGlobalConfiguration: NSObject {

  // Guarded by _loggerLock. nonisolated(unsafe) only silences static-variable isolation
  // checking; the lock is what provides the synchronization.
  nonisolated(unsafe) private static var _logger: (any FBControlCoreLogger)?
  private static let _loggerLock = NSLock()

  // MARK: Timeouts

  @objc public class var fastTimeout: TimeInterval { 10 }
  @objc public class var regularTimeout: TimeInterval { 30 }
  @objc public class var slowTimeout: TimeInterval { 120 }

  // MARK: Logger

  /// The logger used wherever a nullable logger parameter is passed as nil.
  ///
  /// By default this logger is close to silent from a consumer's point of view: it writes to
  /// os_log (subsystem `com.facebook.fbcontrolcore`) at info level and nowhere else. In a process
  /// without a terminal — a GUI app, a launchd job — nothing reaches stderr, so framework
  /// diagnostics (including framework-loading failures) are only visible via
  /// `log stream --predicate 'subsystem == "com.facebook.fbcontrolcore"'`.
  ///
  /// Two environment variables lift this: `FBCONTROLCORE_LOGGING` mirrors output to stderr, and
  /// `FBCONTROLCORE_DEBUG_LOGGING` raises the level to debug. Consumers that want diagnostics
  /// somewhere they will actually look should pass their own logger instead of relying on nil.
  @objc public class var defaultLogger: any FBControlCoreLogger {
    get {
      _loggerLock.lock()
      defer { _loggerLock.unlock() }
      if let existing = _logger { return existing }
      let created = createDefaultLogger()
      _logger = created
      return created
    }
    set {
      _loggerLock.lock()
      let previous = _logger
      _logger = newValue
      _loggerLock.unlock()
      // Outside the critical section: logging through an arbitrary logger implementation must
      // not run under the lock.
      if previous != nil {
        newValue.debug().log("Overriding the Default Logger with \(newValue)")
      }
    }
  }

  // MARK: Configuration

  @objc public class var confirmCodesignaturesAreValid: Bool {
    guard let value = ProcessInfo.processInfo.environment[ConfirmShimsAreSignedEnv] else { return false }
    return (value as NSString).boolValue
  }

  // MARK: NSObject

  override public class func description() -> String {
    _loggerLock.lock()
    let logger = _logger
    _loggerLock.unlock()
    // Stringified outside the critical section: an arbitrary logger implementation must not run
    // under the lock.
    return "Default Logger \(logger.map(String.init(describing:)) ?? "(nil)")"
  }

  public override var description: String {
    Self.description()
  }

  // MARK: Private

  private class func createDefaultLogger() -> any FBControlCoreLogger {
    FBControlCoreLoggerFactory.systemLoggerWriting(toStderr: stderrLoggingEnabledByDefault, withDebugLogging: debugLoggingEnabledByDefault)
  }

  private class var stderrLoggingEnabledByDefault: Bool {
    guard let value = ProcessInfo.processInfo.environment[FBControlCoreStderrLogging] else { return false }
    return (value as NSString).boolValue
  }

  private class var debugLoggingEnabledByDefault: Bool {
    guard let value = ProcessInfo.processInfo.environment[FBControlCoreDebugLogging] else { return false }
    return (value as NSString).boolValue
  }
}
